### PR TITLE
gradient misc optimization and add some helpful function

### PR DESCRIPTION
**Pull request by SantyWang**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Review by zxx43 — APPROVED** [review]

File: cocos/core/curves/gradient.ts [hunk 1]
```diff
@@ -92,24 +94,18 @@ export default class Gradient {
      * @en Array of color key.
      * @zh 颜色关键帧列表。
      */
-    public colorKeys = new Array<ColorKey>();
+    public colorKeys: ColorKey[] = [];
```
Comment (minggo) [inline cocos/core/curves/gradient.ts:97]: Is there any reason to change it?

File: cocos/scene-graph/node.ts [hunk 2]
```diff
@@ -1749,6 +1749,14 @@ export class Node extends CCObject implements ISchedulable, CustomSerializable {
         return this._layer;
     }
 
+    /**
+     * @zh 节点的变换改动版本号。
+     * @en The transformation change version number of the node.
+     */
+    get flagChangedVersion () {
```
Comment (minggo) [inline cocos/scene-graph/node.ts:1758]: Why need this value? As it is used internally, if the value is returned, then can not change to mechanism in future.

File: cocos/core/curves/gradient.ts (hunk 1 above)
Comment (SantyWang) [reply cocos/core/curves/gradient.ts:97]: I think this looks more clearer

File: cocos/scene-graph/node.ts (hunk 2 above)
Comment (SantyWang) [reply cocos/scene-graph/node.ts:1758]: In some scenarios we need to fetch the value and cache it, and compare the current value with the cached value when needed to calculate whether there has been a change in that time.

Comment (minggo) [reply cocos/scene-graph/node.ts:1758]: Could you please describe the scenarios that need it as there is `Node.hasChangedFlags()`;

Comment (SantyWang) [reply cocos/scene-graph/node.ts:1758]: `Node.hasChangedFlags` can only be used to distinguish if there is a change in this frame, if I need to know if a change has occurred in a certain period of time, I need to call Node.hasChangedFlags once for each frame in that period of time. 
With this new interface, I only need to record version once before that time, and again after that time, and compare the following two, if they don't match, then a change has occurred, which is much better and avoids many special cases

Comment (minggo) [reply cocos/scene-graph/node.ts:1758]: Got it, can i know what module need it? As it is implemented depended, so it is better to public it as possible.

Comment (SantyWang) [reply cocos/scene-graph/node.ts:1758]: UI and particle will need it. ui can use it to determine when should transform vertices position into world space. particle use it to recalculate the matrix when the version changed.

Comment (minggo) [reply cocos/scene-graph/node.ts:1758]: If so, it is better to mark it as engine internal.

Comment (SantyWang) [reply cocos/scene-graph/node.ts:1758]: marked as engine internal

**Review by minggo — APPROVED** [review]

**Pull request merged**